# Part I - (Loan Data Exploration From Prosper)
## by Ngwe Becky

## Introduction
> Loan Data from Prosper: This data set contains 113,937 loans with 81 variables on each loan, including loan amount, borrower rate (or interest rate), current loan status, borrower income, and many others.

## Preliminary Wrangling


In [2]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

> Load in your dataset and describe its properties through the questions below. Try and motivate your exploration goals through this section.


In [5]:
# function to read a csv file and save in a pandas dataframe variable
def Upload_csv(dataset):
    data = pd.read_csv(dataset)
    return data

df_loan = Upload_csv('prosperLoanData.csv')
df_loan.head()

ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   
1  10273602499503308B223C1        1209647  2014-02-27 08:28:07.900000000   
2  0EE9337825851032864889A          81716  2007-01-05 15:00:47.090000000   
3  0EF5356002482715299901A         658116  2012-10-22 11:02:35.010000000   
4  0F023589499656230C5E3E2         909464  2013-09-14 18:38:39.097000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   
1         NaN    36    Current                  NaN      0.12016   
2          HR    36  Completed  2009-12-17 00:00:00      0.28269   
3         NaN    36    Current                  NaN      0.12528   
4         NaN    36    Current                  NaN      0.24614   

   BorrowerRate  LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
0        0.1580       0.1380  ...         -133.18                0.0   
1        0.0920       0.0820  ...            0.00                0.0   
2        0.2750       0.2400  ...          -24.20                0.0   
3        0.0974       0.0874  ...         -108.01                0.0   
4        0.2085       0.1985  ...          -60.27                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   
1                    0.0                  0.0                             0.0   
2                    0.0                  0.0                             0.0   
3                    0.0                  0.0                             0.0   
4                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   
1            1.0                0                          0   
2            1.0                0                          0   
3            1.0                0                          0   
4            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  
1                         0.0         1  
2                         0.0        41  
3                         0.0       158  
4                         0.0        20  

[5 rows x 81 columns]

In [6]:
df_loan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

In [7]:
df_loan.describe()

ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.004900        -0.182700   
25%         0.124200                 0.115670       0.042400         0.074080   
50%         0.173000                 0.161500       0.072400         0.091700   
75%         0.240000                 0.224300       0.112000         0.116600   
max         0.492500                 0.319900       0.366000         0.283700   

       ProsperRating (numeric)  ProsperScore  ...  LP_ServiceFees  \
count             84853.000000  84853.000000  ...   113937.000000   
mean                  4.072243      5.950067  ...      -54.725641   
std                   1.673227      2.376501  ...       60.675425   
min                   1.000000      1.000000  ...     -664.870000   
25%                   3.000000      4.000000  ...      -73.180000   
50%                   4.000000      6.000000  ...      -34.440000   
75%                   5.000000      8.000000  ...      -13.920000   
max                   7.000000     11.000000  ...       32.060000   

       LP_CollectionFees  LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
count      113937.000000          113937.000000        113937.000000   
mean          -14.242698             700.446342           681.420499   
std           109.232758            2388.513831          2357.167068   
min         -9274.750000             -94.200000          -954.550000   
25%             0.000000               0.000000             0.000000   
50%             0.000000               0.000000             0.000000   
75%             0.000000               0.000000             0.000000   
max             0.000000           25000.000000         25000.000000   

       LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
count                    113937.000000  113937.000000    113937.000000   
mean                         25.142686       0.998584         0.048027   
std                         275.657937       0.017919         0.332353   
min                           0.000000       0.700000         0.000000   
25%                           0.000000       1.000000         0.000000   
50%                           0.000000       1.000000         0.000000   
75%                           0.000000       1.000000         0.000000   
max                       21117.900000       1.012500        39.000000   

       InvestmentFromFriendsCount  InvestmentFromFriendsAmount      Investors  
count               113937.000000                113937.000000  113937.000000  
mean                     0.023460                    16.550751      80.475228  
std                      0.232412                   294.545422     103.239020  
min                      0.000000                     0.000000       1.000000  
25%                      0.000000                     0.000000       2.000000  
50%                      0.000000                     0.000000      44.000000  
75%                      0.000000                     0.000000     115.000000  
max                     33.000000                 25000.000000    1189.000000

In [8]:
df_loan.sample(10)

ListingKey  ListingNumber            ListingCreationDate  \
9647   032A36007597759893A56A8        1136910  2014-01-14 08:02:47.857000000   
57328  B99035614012895688BE924         666325  2012-11-06 15:33:57.030000000   
95682  F34135926725532839480C2         964723  2013-10-29 13:13:52.610000000   
65418  E5F333972059570289BB3CB         191091  2007-08-22 09:25:01.490000000   
48335  627C3547676180253E346A1         595228  2012-06-01 14:11:56.223000000   
70139  3BFB3486706052653684FE2         461277  2010-06-03 14:09:40.633000000   
48202  9A6D3584286964582CCC2B6         837824  2013-07-13 11:29:58.157000000   
65685  36083503045074903020465         488642  2010-12-21 12:20:56.533000000   
67661  D0813425791334642B0CC3F         371449  2008-07-22 10:59:34.727000000   
32370  82D335961400710574FFDDC        1073729  2013-12-13 19:53:03.830000000   

      CreditGrade  Term            LoanStatus           ClosedDate  \
9647          NaN    60               Current                  NaN   
57328         NaN    36               Current                  NaN   
95682         NaN    36  Past Due (1-15 days)                  NaN   
65418           C    36             Completed  2007-10-11 00:00:00   
48335         NaN    60             Completed  2012-09-04 00:00:00   
70139         NaN    36             Completed  2013-01-25 00:00:00   
48202         NaN    36               Current                  NaN   
65685         NaN    36             Completed  2013-12-11 00:00:00   
67661          AA    36            Chargedoff  2011-06-02 00:00:00   
32370         NaN    36               Current                  NaN   

       BorrowerAPR  BorrowerRate  LenderYield  ...  LP_ServiceFees  \
9647       0.17327        0.1500       0.1400  ...           -9.63   
57328      0.20462        0.1679       0.1579  ...         -157.49   
95682      0.28544        0.2469       0.2369  ...           -9.89   
65418      0.16898        0.1600       0.1450  ...           -1.80   
48335      0.27462        0.2489       0.2389  ...          -25.11   
70139      0.33665        0.3100       0.3000  ...          -25.89   
48202      0.08930        0.0759       0.0659  ...          -70.84   
65685      0.15549        0.1340       0.1240  ...          -58.86   
67661      0.10441        0.0975       0.0875  ...         -100.54   
32370      0.13189        0.1039       0.0939  ...          -12.41   

       LP_CollectionFees  LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
9647                 0.0                    0.0                 0.00   
57328                0.0                    0.0                 0.00   
95682                0.0                    0.0                 0.00   
65418                0.0                    0.0                 0.00   
48335                0.0                    0.0                 0.00   
70139                0.0                    0.0                 0.00   
48202                0.0                    0.0                 0.00   
65685                0.0                    0.0                 0.00   
67661                0.0                 1435.0              1434.99   
32370                0.0                    0.0                 0.00   

      LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
9647                              0.0            1.0                0   
57328                             0.0            1.0                0   
95682                             0.0            1.0                0   
65418                             0.0            1.0                0   
48335                             0.0            1.0                0   
70139                             0.0            1.0                0   
48202                             0.0            1.0                0   
65685                             0.0            1.0                0   
67661                             0.0            1.0                0   
32370                             0.0            1.0                0   



>Note that the dataset consist of 81 columns. For this context, the following columns(variables) were used:

In [9]:
target_columns = [
    'Term', 'LoanStatus', 'BorrowerRate', 'ProsperRating (Alpha)', 'ListingCategory (numeric)', 'EmploymentStatus',
    'DelinquenciesLast7Years', 'StatedMonthlyIncome', 'TotalProsperLoans', 'LoanOriginalAmount',
    'LoanOriginationDate', 'Recommendations', 'Investors'
]

In [10]:
target_df = df_loan[target_columns]

In [11]:
target_df.sample(10)

Term LoanStatus  BorrowerRate ProsperRating (Alpha)  \
25895     36    Current        0.0749                    AA   
103947    36    Current        0.2155                     D   
41220     36    Current        0.1840                     C   
16758     36    Current        0.2452                     D   
105289    60  Completed        0.1490                     A   
27477     36    Current        0.1902                     C   
112701    36    Current        0.0809                    AA   
98441     36  Defaulted        0.3177                    HR   
39313     36    Current        0.0605                    AA   
57117     60    Current        0.2506                     D   

        ListingCategory (numeric) EmploymentStatus  DelinquenciesLast7Years  \
25895                           2         Employed                      0.0   
103947                         13         Employed                      0.0   
41220                           1         Employed                      1.0   
16758                          18         Employed                      0.0   
105289                          1         Employed                      0.0   
27477                           7         Employed                      0.0   
112701                          1         Employed                      0.0   
98441                           7    Self-employed                      0.0   
39313                           1         Employed                      0.0   
57117                           2         Employed                      0.0   

        StatedMonthlyIncome  TotalProsperLoans  LoanOriginalAmount  \
25895           9166.666667                NaN               20000   
103947          5666.666667                NaN                2000   
41220           9000.000000                NaN               15000   
16758           6666.666667                NaN                5000   
105289          3166.666667                NaN               10000   
27477           6163.666667                1.0               15000   
112701          3333.333333                NaN                4916   
98441           2083.333333                NaN                2500   
39313           5000.000000                1.0                5000   
57117          20416.666667                NaN               10000   

        LoanOriginationDate  Recommendations  Investors  
25895   2012-08-28 00:00:00                0        214  
103947  2013-10-16 00:00:00                0         24  
41220   2013-11-04 00:00:00                0          1  
16758   2012-04-06 00:00:00                0         57  
105289  2011-09-08 00:00:00                0        196  
27477   2012-07-02 00:00:00                0        150  
112701  2014-03-03 00:00:00                0          1  
98441   2012-02-03 00:00:00                0          1  
39313   2013-12-16 00:00:00                0          1  
57117   2013-03-27 00:00:00                0          6

In [12]:
target_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 13 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Term                       113937 non-null  int64  
 1   LoanStatus                 113937 non-null  object 
 2   BorrowerRate               113937 non-null  float64
 3   ProsperRating (Alpha)      84853 non-null   object 
 4   ListingCategory (numeric)  113937 non-null  int64  
 5   EmploymentStatus           111682 non-null  object 
 6   DelinquenciesLast7Years    112947 non-null  float64
 7   StatedMonthlyIncome        113937 non-null  float64
 8   TotalProsperLoans          22085 non-null   float64
 9   LoanOriginalAmount         113937 non-null  int64  
 10  LoanOriginationDate        113937 non-null  object 
 11  Recommendations            113937 non-null  int64  
 12  Investors                  113937 non-null  int64  
dtypes: float64(4), int64(5), obje

In [13]:
target_df.describe()

Term   BorrowerRate  ListingCategory (numeric)  \
count  113937.000000  113937.000000              113937.000000   
mean       40.830248       0.192764                   2.774209   
std        10.436212       0.074818                   3.996797   
min        12.000000       0.000000                   0.000000   
25%        36.000000       0.134000                   1.000000   
50%        36.000000       0.184000                   1.000000   
75%        36.000000       0.250000                   3.000000   
max        60.000000       0.497500                  20.000000   

       DelinquenciesLast7Years  StatedMonthlyIncome  TotalProsperLoans  \
count            112947.000000         1.139370e+05       22085.000000   
mean                  4.154984         5.608026e+03           1.421100   
std                  10.160216         7.478497e+03           0.764042   
min                   0.000000         0.000000e+00           0.000000   
25%                   0.000000         3.200333e+03           1.000000   
50%                   0.000000         4.666667e+03           1.000000   
75%                   3.000000         6.825000e+03           2.000000   
max                  99.000000         1.750003e+06           8.000000   

       LoanOriginalAmount  Recommendations      Investors  
count        113937.00000    113937.000000  113937.000000  
mean           8337.01385         0.048027      80.475228  
std            6245.80058         0.332353     103.239020  
min            1000.00000         0.000000       1.000000  
25%            4000.00000         0.000000       2.000000  
50%            6500.00000         0.000000      44.000000  
75%           12000.00000         0.000000     115.000000  
max           35000.00000        39.000000    1189.000000

Since Prosper use their own proprietary Prosper Rating only since 2009, there is a lot of missing values in ProsperRating column. Let's drop these missing values:

In [15]:
target_df = target_df.dropna(subset=['ProsperRating (Alpha)']).reset_index()

Convert LoanOriginationDate to datetime datatype

In [16]:
target_df['LoanOriginationDate']= pd.to_datetime(target_df['LoanOriginationDate'])

Replace the NaN values of TotalProsperLoans with 0

In [18]:
target_df['TotalProsperLoans'] = target_df['TotalProsperLoans'].fillna(0)

In [19]:
target_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84853 entries, 0 to 84852
Data columns (total 15 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   level_0                    84853 non-null  int64         
 1   index                      84853 non-null  int64         
 2   Term                       84853 non-null  int64         
 3   LoanStatus                 84853 non-null  object        
 4   BorrowerRate               84853 non-null  float64       
 5   ProsperRating (Alpha)      84853 non-null  object        
 6   ListingCategory (numeric)  84853 non-null  int64         
 7   EmploymentStatus           84853 non-null  object        
 8   DelinquenciesLast7Years    84853 non-null  float64       
 9   StatedMonthlyIncome        84853 non-null  float64       
 10  TotalProsperLoans          84853 non-null  float64       
 11  LoanOriginalAmount         84853 non-null  int64         
 12  Loan

### What is the structure of your dataset?

> the target_df dataset consist 15 columns and 84853 entries.

### What is/are the main feature(s) of interest in your dataset?
> Trying to figure out what features can be used to predict default on credit. Also i would like to check what are major factors connected with prosper credit rating.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?
> I think that the borrowers Prosper rating will have the highest impact on chances of default. Also I expect that loan amount will play a major role and maybe the category of credit. Prosper rating will depend on stated income and employment status.

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.


> **Rubric Tip**: The project (Parts I alone) should have at least 15 visualizations distributed over univariate, bivariate, and multivariate plots to explore many relationships in the data set.  Use reasoning to justify the flow of the exploration.



>**Rubric Tip**: Use the "Question-Visualization-Observations" framework  throughout the exploration. This framework involves **asking a question from the data, creating a visualization to find answers, and then recording observations after each visualisation.** 




>**Rubric Tip**: Visualizations should depict the data appropriately so that the plots are easily interpretable. You should choose an appropriate plot type, data encodings, and formatting as needed. The formatting may include setting/adding the title, labels, legend, and comments. Also, do not overplot or incorrectly plot ordinal data.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

## Conclusions
>You can write a summary of the main findings and reflect on the steps taken during the data exploration.



> Remove all Tips mentioned above, before you convert this notebook to PDF/HTML


> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML or PDF` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!

